fpga_interchange/arch.cc

```diff
@@ -0,0 +1,598 @@
+    if (pip_info.site != -1) {
+        // This is either a site pin or a site pip.
+        auto &site = chip_info->sites[tile.sites[pip_info.site]];
+        auto &bel = tile_type.bel_data[pip_info.bel];
+        IdString bel_name(bel.name);
+        if(bel.category == BEL_CATEGORY_LOGIC) {
+            // This is a psuedo pip
+            IdString src_wire_name = IdString(tile_type.wire_data[locInfo(pip).pip_data[pip.index].src_index].name);
+            IdString dst_wire_name = IdString(tile_type.wire_data[locInfo(pip).pip_data[pip.index].dst_index].name);
+            return id(site.name.get() + std::string("/") + src_wire_name.str(this) + "." + dst_wire_name.str(this));
+
+        } else if(bel.category == BEL_CATEGORY_ROUTING) {
+            // This is a site pip.
+            IdString pin_name(bel.ports[pip_info.extra_data]);
+            return id(site.name.get() + std::string("/") + bel_name.str(this) + "/" + pin_name.str(this));
+        } else {
+            NPNR_ASSERT(bel.category == BEL_CATEGORY_SITE_PORT);
+            // This is a site pin, just the name of the BEL is a unique identifier.
+            return id(site.name.get() + std::string("/") + bel_name.str(this));
+        }
+    } else {
+        // This is a tile pip.
+        return id(std::string(tile.name.get()) + "/" +
+                  IdString(tile_type.wire_data[locInfo(pip).pip_data[pip.index].src_index].name).str(this) + "." +
+                  IdString(tile_type.wire_data[locInfo(pip).pip_data[pip.index].dst_index].name).str(this));
+    }
+}
+
+IdString Arch::getPipType(PipId pip) const { return id("PIP"); }
+
+std::vector<std::pair<IdString, std::string>> Arch::getPipAttrs(PipId pip) const { return {}; }
+
+// -----------------------------------------------------------------------
+
+std::vector<IdString> Arch::getBelPins(BelId bel) const
+{
+    std::vector<IdString> ret;
+    NPNR_ASSERT(bel != BelId());
+
+    // FIXME: The std::vector here can be replaced by a int32_t -> IdString
+    // range wrapper.
+    int num_bel_wires = locInfo(bel).bel_data[bel.index].num_bel_wires;
+    const int32_t *ports = locInfo(bel).bel_data[bel.index].ports.get();
+
+    for (int i = 0; i < num_bel_wires; i++) {
+        ret.push_back(IdString(ports[i]));
+    }
+
+    return ret;
+}
+
+BelId Arch::getBelByLocation(Loc loc) const
+{
+    BelId bi;
+    if (loc.x >= chip_info->width || loc.y >= chip_info->height)
+        return BelId();
+    bi.tile = getTileIndex(loc);
+    auto &li = locInfo(bi);
+
+    if(loc.z >= li.num_bels) {
+        return BelId();
+    } else {
+        bi.index = loc.z;
+        return bi;
+    }
+}
+
+std::vector<std::pair<IdString, std::string>> Arch::getBelAttrs(BelId bel) const { return {}; }
+
+// -----------------------------------------------------------------------
+
+delay_t Arch::estimateDelay(WireId src, WireId dst, bool debug) const
+{
+    // FIXME: Implement when adding timing-driven place and route.
```

Comment:
estimateDelay is used as an A* heuristic by the router, we will need to return made-up delays instead of 0 for reasonable router behaviour even without timing data.

Reply:
Agreed, I was just trying to get something that would compile. I'll add a TODO so it doesn't get lost.